rapids-cmake/cpm/detail/pinning_write_file.cmake

```diff
@@ -0,0 +1,259 @@
+rapids_cpm_pinning_add_json_entry
+---------------------------------
+
+.. versionadded:: v24.04.00
+
+#]=======================================================================]
+function(rapids_cpm_pinning_add_json_entry json_var package_name url_var sha_var)
+  include("${rapids-cmake-dir}/cpm/detail/get_default_json.cmake")
+  include("${rapids-cmake-dir}/cpm/detail/get_override_json.cmake")
+  get_default_json(${package_name} json_data)
+  get_override_json(${package_name} override_json_data)
+
+  set(url_string)
+  set(sha_string)
+  if(${url_var})
+    string(CONFIGURE [=["git_url": "${${url_var}}",]=] url_string)
+  endif()
+  if(${sha_var})
+    string(CONFIGURE [=["git_tag": "${${sha_var}}",]=] sha_string)
+  endif()
+  # We start with a default template, and only add members that don't exist
+  string(CONFIGURE [=[{
+  "version": "${CPM_PACKAGE_${package_name}_VERSION}",
+  ${url_string}
+  ${sha_string}
+  "git_shallow": false,
+  "always_download": true
+  }]=]
+                   pinned_json_entry)
+
+  # Insert a json key and value
+  #
+  function(rapids_cpm_pinning_create_and_set_member json_blob_var key value)
+
+    # Identify special values types that shouldn't be treated as a string
+    # https://gitlab.kitware.com/cmake/cmake/-/issues/25716
+    if(value MATCHES "(^true$|^false$|^null$|^\\{|^\\[)")
+      # value is a json type that doesn't need quotes
+      string(JSON json_blob ERROR_VARIABLE err_var SET "${${json_blob_var}}" ${key} ${value})
+    else()
+      # We need to quote 'value' so that it is a valid string json element.
+      string(JSON json_blob ERROR_VARIABLE err_var SET "${${json_blob_var}}" ${key} "\"${value}\"")
+    endif()
+    set(${json_blob_var} "${json_blob}" PARENT_SCOPE)
+
+  endfunction()
+
+  foreach(data IN LISTS override_json_data json_data)
+    if(NOT data)
+      # Need to handle both json_data and the override being empty
+      continue()
+    endif()
+    string(JSON entry_count LENGTH "${data}")
+    math(EXPR entry_count "${entry_count} - 1")
+    # cmake-lint: disable=E1120
+    foreach(index RANGE ${entry_count})
+      string(JSON member MEMBER "${data}" ${index})
+      string(JSON existing_value ERROR_VARIABLE dont_have GET "${pinned_json_entry}" ${member})
+      if(dont_have)
+        string(JSON value GET "${data}" ${member})
+        rapids_cpm_pinning_create_and_set_member(pinned_json_entry ${member} ${value})
+      endif()
+    endforeach()
+  endforeach()
+  set(${json_var} "\"${package_name}\": ${pinned_json_entry}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_pretty_format_json
+-------------------------------------
+
+
+.. versionadded:: v24.04.00
+
+Formats provided JSON to be easily read by humans
+#]=======================================================================]
+function(rapids_cpm_pinning_pretty_format_json _rapids_json_var)
```

Comment:
Since we're constructing our JSON string why can't we just construct it to be pretty from the start so that we don't need to prettify it afterwards?

Reply:
Each package fragment constructed in `rapids_cpm_pinning_add_json_entry` is considered to be a root element when we are using it. This means that the internal JSON engine to CMake will format it to an indentation level that is incorrect once we wrap them together in the `package` object.

What I can look at doing is seeing if we can get cmake to reformat everything for us, using a failed `SET` call or something instead of having a custom formatting engine.

Reply:
The formatting that CMake would generate would be:
```
{
  "packages" : 
  {
    "CCCL" : 
    {
      "always_download" : true,
      "git_shallow" : false,
      "git_tag" : "36f379f29660761fe033a1306ca9dab6a88cb65c",
      "git_url" : "https://github.com/NVIDIA/cccl.git",
      "patches" : 
      [
....
```

Are we happy with that style of JSON compared to the more compact version that we have in `versions.json`. Personally I am happy with it since it would remove the need to have a custom ( and hacky ) pretty printer.

Reply:
That’s fine. I don’t like the format but prefer not having the extra code for pretty print.

Reply:
Yeah I'm in the same boat, I don't think the improved formatting is worth the extra code.